### PR TITLE
An ns-3 simulated network now sits between endpoints.

### PR DESCRIPTION
**Pull request by xvzcf**
The plan next is to add packet decryption and log capturing functionality. I'll refactor this over the course of that work.

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -1,6 +1,25 @@
...
+    hostname: network
+    stdin_open: true
+    tty: true
+    cap_add:
+      - NET_ADMIN
+    expose:
+      - "57832"
+    networks:
+      clientnet:
+        ipv4_address: 193.167.0.2
```
Comment (cjpatton) [inline docker-compose.yml:17]: I'm curious how the network 193.167.0.0/16 was chosen?

File: impl-endpoints/cloudflare-go/setup.sh [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
```
Comment (cjpatton) [inline impl-endpoints/cloudflare-go/setup.sh:1]: This looks to be the same script as `impl-endpoints/boringssl/setup.sh`. Any way to deduplicate this?

File: impl-endpoints/rustls/setup.sh [hunk 3]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
```
Comment (cjpatton) [inline impl-endpoints/rustls/setup.sh:1]: Same comment as above.

**Review by cjpatton — APPROVED** [review]
I'm agnostic about using NS-3, so I simply checked that this runs on my machine. Approved, but I would suggest you deduplicate the `setup.sh` scripts so that this doesn't become a maintenance headache later on.

File: docker-compose.yml (hunk 1 above)
Comment (xvzcf) [reply docker-compose.yml:17]: I just picked the IP ranges used in the [QUIC Interop Runner](https://github.com/marten-seemann/quic-interop-runner/blob/master/docker-compose.yml#L19)

**Comment by xvzcf** [comment]
I agree that `setup.sh` should be de-duplicated. It's on the TODO list, I'm trying to figure out if I can specify a gateway in the docker-compose file and/or possibly apply some additional settings just to the ns-3 network image.

**Comment by cjpatton** [comment]
> I agree that `setup.sh` should be de-duplicated. It's on the TODO list, I'm trying to figure out if I can specify a gateway in the docker-compose file and/or possibly apply some additional settings just to the ns-3 network image.

I'm out of my depth, so I'll leave you to it :) Although I'm sure it would be helpful for others to add "TODO(xvzcf)" comments in the code in places where you intend to fix things / make improvements.

**Comment by xvzcf** [comment]
> I'm out of my depth, so I'll leave you to it :) Although I'm sure it would be helpful for others to add "TODO(xvzcf)" comments in the code in places where you intend to fix things / make improvements.

Good point, will do.

**Pull request merged**